### PR TITLE
feat(adcn): introduce the avionics data communication network

### PR DESCRIPTION
**Pull request by svengcz**
## Summary of Changes

Introduces the ADCN system of the A380.
The CPIOMs, IOMs and the AFDX networks are implemented.

A routing table is calculated which provides the information which AFDX switch is reachable for every other switch.
A component needs to know it's connection to the dedicated AFDX switch and a call in the look-up table gives the information if the remote system is reachable.

## Screenshots (if necessary)

ADCN routes

## Additional context

Discord username (if different from GitHub):

## Testing instructions

No QA tests required

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: src/systems/systems/src/integrated_modular_avionics/avionics_full_duplex_switch.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+pub struct AvionicsFullDuplexSwitch {
+    single_power_supply: Option<ElectricalBusType>,
+    power_supply_relay: Option<PowerSupplyRelay>,
+    last_is_powered: bool,
+    is_powered: bool,
+    failure_indication_id: VariableIdentifier,
+    last_failure_indication: bool,
+    failure_indication: bool,
+    available_id: VariableIdentifier,
```
Comment (2hwk) [inline src/systems/systems/src/integrated_modular_avionics/avionics_full_duplex_switch.rs:18]: see rust guidelines.md and readme.md ideally systems should not be aware they are running in a simulator and simvar state should be passed and vars not being read here i.e. the system is unaware that it is running in a simulator

**Review by 2hwk — CHANGES_REQUESTED** [review]

File: src/systems/systems/src/integrated_modular_avionics/avionics_full_duplex_switch.rs (hunk 1 above)
Comment (2hwk) [reply src/systems/systems/src/integrated_modular_avionics/avionics_full_duplex_switch.rs:18]: Resolved in development channel on discord

File: src/systems/systems/src/integrated_modular_avionics/core_processing_input_output_module.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        let failure_value: f64 = reader.read(&self.failure_indication_id);
+        self.failure_indication = failure_value != 0.0;
+    }
+
+    fn write(&self, writer: &mut SimulatorWriter) {
+        if self.is_available() {
+            writer.write(&self.available_id, 1.0);
+        } else {
+            writer.write(&self.available_id, 0.0);
+        }
```
Comment (2hwk) [inline src/systems/systems/src/integrated_modular_avionics/core_processing_input_output_module.rs:44]: possibility of simple assignment or tenery to save the use of if else? (not that familiar with rs sorry)

File: src/systems/systems/src/integrated_modular_avionics/input_output_module.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        let failure_value: f64 = reader.read(&self.failure_indication_id);
+        self.failure_indication = failure_value != 0.0;
+    }
+
+    fn write(&self, writer: &mut SimulatorWriter) {
+        if self.is_available() {
+            writer.write(&self.available_id, 1.0);
+        } else {
+            writer.write(&self.available_id, 0.0);
+        }
```
Comment (2hwk) [inline src/systems/systems/src/integrated_modular_avionics/input_output_module.rs:44]: ditto

**Review by 2hwk — CHANGES_REQUESTED** [review]

File: src/systems/systems/src/integrated_modular_avionics/core_processing_input_output_module.rs (hunk 2 above)
Comment (Gurgel100) [reply src/systems/systems/src/integrated_modular_avionics/core_processing_input_output_module.rs:44]: I would actually recommend doing it like this:
```suggestion
        writer.write(&self.available_id, (self.is_available() as u64) as f64);
```

Comment (2hwk) [reply src/systems/systems/src/integrated_modular_avionics/core_processing_input_output_module.rs:44]: actually this might be possible 

```suggestion
      writer.write(&self.available_id, self.is_available());
```

Comment (svengcz) [reply src/systems/systems/src/integrated_modular_avionics/core_processing_input_output_module.rs:44]: fixed

File: src/systems/systems/src/integrated_modular_avionics/input_output_module.rs (hunk 3 above)
Comment (svengcz) [reply src/systems/systems/src/integrated_modular_avionics/input_output_module.rs:44]: fixed

File: src/systems/systems/src/integrated_modular_avionics/core_processing_input_output_module.rs (hunk 2 above)
Comment (svengcz) [reply src/systems/systems/src/integrated_modular_avionics/core_processing_input_output_module.rs]: fixed

File: fbw-common/src/wasm/systems/systems/src/integrated_modular_avionics/avionics_full_duplex_switch.rs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    pub fn routing_update_required(&self) -> bool {
+        self.routing_update_required
+    }
+}
+
+impl SimulationElement for AvionicsFullDuplexSwitch {
+    fn accept<T: SimulationElementVisitor>(&mut self, visitor: &mut T) {
+        if self.power_supply_relay.is_some() {
+            self.power_supply_relay.as_mut().unwrap().accept(visitor);
+        }
```
Comment (Gurgel100) [inline fbw-common/src/wasm/systems/systems/src/integrated_modular_avionics/avionics_full_duplex_switch.rs:93]: If you don't use the enum:
```suggestion
        if let Some(ref mut power_supply_relay) = self.power_supply {
            power_supply_relay.accept(visitor);
        }
```

File: fbw-common/src/wasm/systems/systems/src/integrated_modular_avionics/avionics_full_duplex_switch.rs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            writer.write(&self.available_id, 1.0);
+        } else {
+            writer.write(&self.available_id, 0.0);
+        }
+    }
+
+    fn receive_power(&mut self, buses: &impl ElectricalBuses) {
+        if self.single_power_supply.is_some() {
+            self.is_powered = buses.is_powered(self.single_power_supply.unwrap());
+        }
```
Comment (Gurgel100) [inline fbw-common/src/wasm/systems/systems/src/integrated_modular_avionics/avionics_full_duplex_switch.rs:113]: If you don't use the enum:
```suggestion
        if let Some(single_power_supply) = self.single_power_supply {
            self.is_powered = buses.is_powered(single_power_supply);
        }
```

File: fbw-common/src/wasm/systems/systems/src/integrated_modular_avionics/core_processing_input_output_module.rs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    pub fn is_available(&self) -> bool {
+        self.is_powered & !self.failure_indication
+    }
+}
+
+impl SimulationElement for CoreProcessingInputOutputModule {
+    fn read(&mut self, reader: &mut SimulatorReader) {
+        let failure_value: f64 = reader.read(&self.failure_indication_id);
+        self.failure_indication = failure_value != 0.0;
```
Comment (Gurgel100) [inline fbw-common/src/wasm/systems/systems/src/integrated_modular_avionics/core_processing_input_output_module.rs:36]:
```suggestion
        self.failure_indication = reader.read(&self.failure_indication_id);
```

File: fbw-common/src/wasm/systems/systems/src/integrated_modular_avionics/input_output_module.rs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    pub fn is_available(&self) -> bool {
+        self.is_powered & !self.failure_indication
+    }
+}
+
+impl SimulationElement for InputOutputModule {
+    fn read(&mut self, reader: &mut SimulatorReader) {
+        let failure_value: f64 = reader.read(&self.failure_indication_id);
+        self.failure_indication = failure_value != 0.0;
```
Comment (Gurgel100) [inline fbw-common/src/wasm/systems/systems/src/integrated_modular_avionics/input_output_module.rs:36]:
```suggestion
        self.failure_indication = reader.read(&self.failure_indication_id);
```

File: fbw-common/src/wasm/systems/systems/src/shared/mod.rs [hunk 8]
```diff
@@ -16,14 +16,14 @@ use uom::si::{
 
 pub mod low_pass_filter;
 pub mod pid;
+pub mod power_supply_relay;
```
Comment (Gurgel100) [inline fbw-common/src/wasm/systems/systems/src/shared/mod.rs:19]: Why did you change this?

**Review by Gurgel100 — CHANGES_REQUESTED** [review]
This took some time ;). Hopefully not too overhelming.

File: fbw-a380x/src/wasm/systems/a380_systems/src/avionics_data_communication_network.rs [hunk 9]
```diff
@@ -0,0 +1,760 @@
...
+
+struct RoutingTableEntry {
+    routing_id_1: VariableIdentifier,
+    routing_id_2: VariableIdentifier,
+    reachable: bool,
+}
+
+/*
+ * The routing table entry describes if two AFDX switches are reachable
+ */
```
[the remark this answers is not in the document]
Comment (svengcz) [reply fbw-a380x/src/wasm/systems/a380_systems/src/avionics_data_communication_network.rs:25]: fixed

File: fbw-a380x/src/wasm/systems/a380_systems/src/avionics_data_communication_network.rs [hunk 10]
```diff
@@ -0,0 +1,760 @@
...
+                    17,
+                    ElectricalBusType::DirectCurrent(1),
+                ),
+                AvionicsFullDuplexSwitch::new_dual_power_supply(
+                    context,
+                    19,
+                    ElectricalBusType::DirectCurrent(1),
+                    ElectricalBusType::DirectCurrentEssential,
+                ),
+            ],
```
[the remark this answers is not in the document]
Comment (svengcz) [reply fbw-a380x/src/wasm/systems/a380_systems/src/avionics_data_communication_network.rs:158]: Yes, 8, 10 and 18 are missing. It is how the network is designed

File: fbw-common/src/wasm/systems/systems/src/shared/mod.rs (hunk 8 above)
Comment (svengcz) [reply fbw-common/src/wasm/systems/systems/src/shared/mod.rs:19]: no, was a merge error

**Comment by svengcz** [comment]
> This took some time ;). Hopefully not too overhelming.

Thanks for the great review!
I've learned a lot about rust.

All suggestions are integrated

File: fbw-a380x/src/wasm/systems/a380_systems/src/avionics_data_communication_network.rs [hunk 11]
```diff
@@ -610,16 +515,14 @@ mod tests {
...
-                ));
-                let reachable_second: f64 = test_bed.read_by_name(Box::leak(
-                    format!("AFDX_{}_{}_REACHABLE", switch, fixed_id).into_boxed_str(),
-                ));
+                let reachable_first: bool =
+                    test_bed.read_by_name(&format!("AFDX_{}_{}_REACHABLE", fixed_id, switch));
+                let reachable_second: bool =
+                    test_bed.read_by_name(&format!("AFDX_{}_{}_REACHABLE", switch, fixed_id));
 
                 println!("AFDX switch combination: {} {}", fixed_id, switch);
```
Comment (Gurgel100) [inline fbw-a380x/src/wasm/systems/a380_systems/src/avionics_data_communication_network.rs:523]: Remove print

File: fbw-common/src/wasm/systems/systems/src/integrated_modular_avionics/avionics_full_duplex_switch.rs [hunk 12]
```diff
@@ -5,11 +5,15 @@ use crate::{
         SimulatorWriter, VariableIdentifier, Write,
     },
 };
-use std::option::Option;
+// use std::option::Option;
```
Comment (Gurgel100) [inline fbw-common/src/wasm/systems/systems/src/integrated_modular_avionics/avionics_full_duplex_switch.rs:8]: Remove unused code

**Review by Gurgel100 — CHANGES_REQUESTED** [review]
Two last things

**Review by Gurgel100 — APPROVED** [review]
LGTM. Great work

**Review by 2hwk — APPROVED** [review]
